core/common/src/main/scala/org/typelevel/otel4s/context/propagation/PassThroughPropagator.scala

```diff
@@ -0,0 +1,79 @@
+/*
+ * Copyright 2022 Typelevel
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.typelevel.otel4s.context
+package propagation
+
+import cats.effect.SyncIO
+import org.typelevel.otel4s.context.syntax._
+
+/** A [[TextMapPropagator]] that extracts a specified collection of fields and
+  * stores them in a context, and extracts them from a context later for
+  * injection. It does not interact with telemetry.
+  */
+final class PassThroughPropagator[Ctx, K[X] <: Key[X]] private (
+    val fields: Iterable[String],
+    entriesKey: K[List[(String, String)]]
+)(implicit c: Contextual.Keyed[Ctx, K])
+    extends TextMapPropagator[Ctx] {
+
+  def extract[A](ctx: Ctx, carrier: A)(implicit
+      getter: TextMapGetter[A]
+  ): Ctx = {
+    val list = fields.view
+      .flatMap(k => getter.get(carrier, k).map(k -> _))
+      .toList
+    if (list.isEmpty) ctx else ctx.updated(entriesKey, list)
+  }
+
+  def inject[A](ctx: Ctx, carrier: A)(implicit updater: TextMapUpdater[A]): A =
+    ctx
+      .getOrElse(entriesKey, Nil)
+      .foldLeft(carrier) { case (c, k -> v) => updater.updated(c, k, v) }
+
+  override def toString: String = s"PassThroughPropagator{fields=$fields}"
+}
+
+object PassThroughPropagator {
+  private def forDistinctFields[Ctx, K[X] <: Key[X]](fields: Seq[String])(
+      implicit
+      c: Contextual.Keyed[Ctx, K],
+      kp: Key.Provider[SyncIO, K]
+  ): TextMapPropagator[Ctx] =
+    if (fields.isEmpty) TextMapPropagator.noop
+    else {
+      new PassThroughPropagator(
+        fields,
+        kp.uniqueKey[List[(String, String)]](
+          "otel4s-PassThroughPropagator-entries"
+        ).unsafeRunSync()
+      )
+    }
+
+  /** Creates a `PassThroughPropagator` that propagates the given fields. */
+  def apply[Ctx, K[X] <: Key[X]](fields: String*)(implicit
+      c: Contextual.Keyed[Ctx, K],
+      kp: Key.Provider[SyncIO, K]
+  ): TextMapPropagator[Ctx] =
+    forDistinctFields(fields.distinct)
+
+  /** Creates a `PassThroughPropagator` that propagates the given fields. */
+  def apply[Ctx, K[X] <: Key[X]](fields: IterableOnce[String])(implicit
```

Comment:
I'm not sure about `IterableOnce`. In my opinion, it's an error-prone type, because some entries may be lost:

```scala
// unfolds the following values: "3", "2", "1"
val it = Iterator.unfold(3) { count =>
  Option.when(count > 0)((count.toString, count - 1))
}

it.next() // returns 3

PassTroughPropagator(it) // compiles fine. However, you will get only `2` and `1`
```

So, I guess we should stick with `Iterable[String]`. It's broad enough to support the variety of subtypes.

Reply:
I disagree with this for a few reasons:

1. you can shoot yourself in the foot just as much with mutable collections, which extend `Iterable` and would still be permitted by your suggested signature.
2. while `Iterator` is a subtype of `IterableOnce`, it is by no means the only one. `IterableOnce` also includes things that can be iterated over (even arbitrarily many times), but which aren't quite collections for a variety of reasons. for example, `Option` is `IterableOnce`, but is not `Iterable` because you cannot use `Option` as a collection of arbitrary elements (i.e. it has no meaningful factory or builder).
3. we only need to iterate over the fields once because we have to make sure they are distinct and because we convert them into a `Seq`, so `IterableOnce` is the highest up type in the collections hierarchy that is sufficient for our needs.

admittedly, this is not the first time I've clashed with other folks in the cats/pure FP community about what collection types to use, and is unlikely to be the last either

Reply:
> which extend Iterable

To be sure, we can use `scala.collection.immutable.Iterable`.

> we only need to iterate over the fields 

I agree. The potential problem is that we can lose information in _some_ cases:
```scala
val it: IterableOnce[String] = Iterator.unfold(3)(count => Option.when(count > 0)((count.toString, count - 1)))
val propagator1 = PassThroughPropagator(it) // 
propagator1.fields // has "3", "2", "1"
val propagator2 = PassThroughPropagator(it) 
propagator2.fields // is empty
```

I agree that that's an unlikely scenario. But with `scala.collection.immutable.Iterable` it should be impossible.

Reply:
to be consistent, I think I'll use `Foldable` as was suggested on... #382?

Reply:
... hmm. it's surprisingly hard to use `Foldable` with `distinct`

Reply:
`Foldable` does not support the current tests, which is a mark against it IMO. I've changed it to `Iterable`